### PR TITLE
Tmi2 365/build mandatory question first page and service layer

### PR DESCRIPTION
**Pull request by a-lor-cab**
## Description

[TMI2-365](https://technologyprogramme.atlassian.net/browse/TMI2-365)

Summary of the changes and the related issue. List any dependencies that are required for this change:
add mandatory-question name page with getServerSide props
add service that will be used by getServerSideProps
add skeleton for other needed pages
add routes to util
add util to determine if a page has already been answered.
add unit test coverage

## Type of change

Please check the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [√] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes:

- [√] Unit Test

- [ ] Integration Test (if applicable)

- [ ] End to End Test (if applicable)

## Screenshots (if appropriate):

Please attach screenshots of the change if it is a UI change:

# Checklist:

- [ ] If I have listed depedenencies above, I have ensured that they are present in the target branch.
- [ ] I have performed a self-review of my code.
- [ ] I have commented my code in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation where applicable.
- [ ] I have ran cypress tests and they all pass locally.

[TMI2-365]: https://technologyprogramme.atlassian.net/browse/TMI2-365?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/applicant/src/pages/mandatory-questions/[mandatoryQuestionId]/getServerSideProps.test.tsx [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+    companiesHouseNumber: null,
+    orgType: null,
+    fundingAmount: null,
+    fundingLocation: null,
+  });
+
+  beforeEach(() => {
+    jest.resetAllMocks();
+    process.env.USER_TOKEN_NAME = 'gap-test';
+  });
```
Comment (ConorFayleTCO) [inline packages/applicant/src/pages/mandatory-questions/[mandatoryQuestionId]/getServerSideProps.test.tsx:53]: could set up a beforeEach and afterEach to reset the env var to its original value

like
```
  let userTokenNameBackup;
  beforeEach(() => {
    jest.resetAllMocks();
    userTokenNameBackup = process.env.USER_TOKEN_NAME
    process.env.USER_TOKEN_NAME = 'gap-test';
  });
  afterEach(() => {
    process.env.USER_TOKEN_NAME = userTokenNameBackup;
  });
```

**Review by ConorFayleTCO — DISMISSED** [review]

File: packages/applicant/src/pages/mandatory-questions/[mandatoryQuestionId]/organisation-name.page.tsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+              fieldErrors={fieldErrors}
+              width="30"
+            />
+
+            <SaveAndCancel
+              type={ButtonTypePropertyEnum.Submit}
+              saveButton={{
+                name: 'Save and continue',
+              }}
+            />
```
Comment (ConorFayleTCO) [inline packages/applicant/src/pages/mandatory-questions/[mandatoryQuestionId]/organisation-name.page.tsx:53]: don't think we need a Cancel button so should just be
```suggestion
            <Button
              text="Save and continue"
              type={ButtonTypePropertyEnum.Submit}
            />
```

File: packages/applicant/src/pages/api/create-mandatory-question.page.tsx [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import { NextApiRequest, NextApiResponse } from 'next';
+import { GrantMandatoryQuestionService } from '../../services/GrantMandatoryQuestionService';
+import { getJwtFromCookies } from '../../utils/jwt';
+import { routes } from '../../utils/routes';
+export default async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse
+) {
+  const schemeId = req.query.schemeId as string;
+  console.log('schemeId', schemeId);
```
Comment (ConorFayleTCO) [inline packages/applicant/src/pages/api/create-mandatory-question.page.tsx:10]: remove

File: packages/applicant/src/pages/api/create-mandatory-question.page.tsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  try {
+    const grantMandatoryQuestionService =
+      GrantMandatoryQuestionService.getInstance();
+    console.log('grantMandatoryQuestionService', grantMandatoryQuestionService);
+    console.log(
+      'BANANA',
```
Comment (ConorFayleTCO) [inline packages/applicant/src/pages/api/create-mandatory-question.page.tsx:16]: lol

File: packages/applicant/src/pages/mandatory-questions/[mandatoryQuestionId]/getServerSideProps.tsx [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  let defaultFields = mandatoryQuestion as Optional<GrantMandatoryQuestionDto>;
+  let fieldErrors = [] as ValidationError[];
+
+  //if there are validation errors we fill up the default fields with the values that the user has entered, and field errors with the errors
+  if ('fieldErrors' in response) {
+    fieldErrors = response.fieldErrors;
+    defaultFields = response.body as Optional<GrantMandatoryQuestionDto>;
+  }
```
Comment (iaincooper-tco) [inline packages/applicant/src/pages/mandatory-questions/[mandatoryQuestionId]/getServerSideProps.tsx:112]: not sure on this, that Optional is a unit test helper and we're using in live code.

File: packages/applicant/src/utils/mandatoryQuestionUtils.ts [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    mandatoryQuestion.addressLine1 !== null &&
+    mandatoryQuestion.city !== null &&
+    mandatoryQuestion.postcode !== null
+  ) {
+    return routes.mandatoryQuestions.typePage(mandatoryQuestionId);
+  }
+  if (mandatoryQuestion.name !== null) {
+    return routes.mandatoryQuestions.addressPage(mandatoryQuestionId);
+  }
+};
```
Comment (iaincooper-tco) [inline packages/applicant/src/utils/mandatoryQuestionUtils.ts:78]: interesting approach, I can see why you've done it as it keeps things common and gives us an entire structure journey (which explains a lot about the size of this).

**Review by iaincooper-tco — CHANGES_REQUESTED** [review]

File: packages/applicant/src/pages/api/create-mandatory-question.page.tsx [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+      await grantMandatoryQuestionService.createMandatoryQuestion(
+        schemeId,
+        getJwtFromCookies(req)
+      )
+    );
+    const mandatoryQuestionId =
+      await grantMandatoryQuestionService.createMandatoryQuestion(
+        schemeId,
+        getJwtFromCookies(req)
+      );
```
[the remark this answers is not in the document]
Comment (a-lor-cab) [reply packages/applicant/src/pages/api/create-mandatory-question.page.tsx:26]: it was a console log, above. forgot to remove it

File: packages/applicant/src/pages/api/create-mandatory-question.page.tsx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+    res.redirect(
+      `${process.env.HOST}${routes.mandatoryQuestions.namePage(
+        mandatoryQuestionId
+      )}`
+    );
+  } catch (e) {
+    console.log('e', e);
```
Comment (iaincooper-tco) [inline packages/applicant/src/pages/api/create-mandatory-question.page.tsx:26]: console.error

**Review by iaincooper-tco — DISMISSED** [review]

**Review by iaincooper-tco — APPROVED** [review]

**Pull request merged**